rocketpy/rocket/rocket.py

```diff
@@ -419,47 +445,79 @@ def evaluate_thrust_to_weight(self):
-            Float value corresponding to rocket static margin when
-            loaded with propellant in units of rocket diameter or
-            calibers.
+        self.cp_position : Function
+            Function of Mach number expressing the rocket's center of pressure
+            position relative to user defined rocket reference system.
+            See :doc:`Positions and Coordinate Systems </user/positions>`
+            for more information.
         """
-        # Initialize total lift coefficient derivative and center of pressure position
-        self.total_lift_coeff_der = 0
-        self.cp_position = 0
+        # Re-Initialize total lift coefficient derivative and center of pressure position
+        self.total_lift_coeff_der.set_source(lambda mach: 0)
+        self.cp_position.set_source(lambda mach: 0)
 
         # Calculate total lift coefficient derivative and center of pressure
         if len(self.aerodynamic_surfaces) > 0:
             for aero_surface, position in self.aerodynamic_surfaces:
-                self.total_lift_coeff_der += aero_surface.clalpha(0)
-                self.cp_position += aero_surface.clalpha(0) * (
+                self.total_lift_coeff_der += aero_surface.clalpha
+                self.cp_position += aero_surface.clalpha * (
                     position - self._csys * aero_surface.cpz
                 )
             self.cp_position /= self.total_lift_coeff_der
 
+        return self.cp_position
+
+    def evaluate_stability_margin(self):
+        """Calculates the stability margin of the rocket as a function of mach
+        number and time.
+
+        Returns
+        -------
+        stability_margin : Function
+            Stability margin of the rocket, in calibers, as a function of mach
+            number and time. Stability margin is defined as the distance between
+            the center of pressure and the center of mass, divided by the
+            rocket's diameter.
+        """
+        self.stability_margin.set_source(
+            lambda mach, time: (
+                (self.center_of_mass(time) - self.cp_position(mach)) / (2 * self.radius)
+            )
+            * self._csys
+        )
+        return self.stability_margin
+
+    def evaluate_static_margin(self):
+        """Calculates the static margin of the rocket as a function of time.
+
+        Returns
+        -------
+        static_margin : Function
+            Static margin of the rocket, in calibers, as a function of time.
+            Static margin is defined as the distance between the center of
+            pressure and the center of mass, divided by the rocket's diameter.
+        """
         # Calculate static margin
-        self.static_margin = (self.center_of_mass - self.cp_position) / (
-            2 * self.radius
+        self.static_margin.set_source(
+            lambda time: (self.center_of_mass(time) - self.cp_position(0))
+            / (2 * self.radius)
+            * self._csys
```

Comment:
I have no problems with this, but out of curiosity why define a `lambda` source and discretize right after?

Wouldn't it work using `Function` operations directly? (same comment to the `stability_margin` evaluation)

Reply:
For some reason, static_margin used to break pretty badly with some liquid and hybrid motors. We added this discretization when developing the liquid and hybrid motors, but it might have been done before we started discretizing things in those classes

I just tested it and it seems it can be remover without issues